File: src/test/kotlin/sigma/software/leovegas/drugstore/service/ProductServiceTest.java
package sigma.software.leovegas.drugstore.service;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import sigma.software.leovegas.drugstore.dto.ProductDto;
import sigma.software.leovegas.drugstore.exception.ResourceNotFoundException;
import sigma.software.leovegas.drugstore.persistence.entity.Product;
import sigma.software.leovegas.drugstore.persistence.repository.ProductRepository;

import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
public class ProductServiceTest {
    BigDecimal bigDecimal = new BigDecimal("22.5");
    ProductDto productDto = ProductDto.builder()
            .name("my")
            .quantity(2)
            .price(bigDecimal)
            .build();
    Product product = Product.builder()
            .name("my")
            .quantity(2)
            .price(bigDecimal)
            .build();

    @InjectMocks
    private ProductServiceImpl service;

    @Mock
    private ProductRepository repo;

    @Test
    public void should_be_created_new_Product() {
        when(repo.save(product)).thenReturn(product);

        ProductDto actual = service.create(productDto);
        Assertions.assertEquals(productDto, actual);
    }

    @Test
    public void IfSaveEmptyProd_shouldReturnException() {
        when(repo.save(null)).thenThrow(NullPointerException.class);

        Assertions.assertThrows(NullPointerException.class, () -> {
            repo.save(null);
        });
    }

    @Test
    public void should_return_All_products() {
        List<Product> products = new ArrayList<>();
        products.add(product);
        List<ProductDto> productsDto = new ArrayList<>();
        productsDto.add(productDto);
        when(repo.findAll()).thenReturn(products);

        List<ProductDto> actual = service.getAll();
        Assertions.assertEquals(productsDto, actual);
    }

    @Test
    public void ifExist_should_returnOneProduct() {
        Long existID = 12L;
        when(repo.findById(existID)).thenReturn(java.util.Optional.ofNullable(product));

        ProductDto actual = service.getOne(existID);
        Assertions.assertEquals(productDto, actual);
    }

    @Test
    public void ifNotExist_shouldReturn_ResourceNotFoundException(){
        Long notExistID = 20L;
        when(repo.findById(notExistID)).thenThrow(ResourceNotFoundException.class);

        Assertions.assertThrows(ResourceNotFoundException.class, () -> {
            service.getOne(notExistID);
        });
    }

}

File: src/test/kotlin/sigma/software/leovegas/drugstore/service/ProductServiceTest.kt
package sigma.software.leovegas.drugstore.service

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase
import org.springframework.boot.test.context.SpringBootTest
import sigma.software.leovegas.drugstore.dto.ProductRequest
import sigma.software.leovegas.drugstore.dto.ProductResponse
import sigma.software.leovegas.drugstore.exception.ResourceNotFoundException
import java.math.BigDecimal


@SpringBootTest
@AutoConfigureTestDatabase
class ProductKotlinServiceTest() {
    val productRequest = ProductRequest(name = "test", price = BigDecimal("25.5"), quantity = 5)
    val productResponse = ProductResponse(name = "test", price = BigDecimal("25.5"), quantity = 5)

    @Autowired
    lateinit var service: ProductServiceImpl

    @Test
    fun `get all should return List of productsResponse`() {
        val all = service.getAll()

        Assertions.assertNotNull(all)
        Assertions.assertTrue(all::class.qualifiedName == "java.util.ArrayList")
    }

    @Test
    fun `if exist should return productDto`() {
        val actual = service.getOne(1)

        Assertions.assertNotNull(actual)
        Assertions.assertTrue(actual.javaClass.kotlin == ProductResponse::class)
    }

    @Test
    fun `if not exist should return ResourceNotFoundException`() {
        assertThrows<ResourceNotFoundException> {
            service.getOne(Long.MAX_VALUE)
        }
    }

    @Test
    fun `when create should return productDto`() {
        val actual = service.save(productRequest)

        Assertions.assertNotNull(actual)
        Assertions.assertNotNull(actual.id)
        Assertions.assertEquals(productResponse.name, actual.name)
        Assertions.assertEquals(productResponse.price, actual.price)
        Assertions.assertEquals(productResponse.quantity, actual.quantity)
    }

    @Test
    fun `when update should return updated productDto`() {
        val randomName = Math.random().toString()
        productRequest.name = randomName
        val actual = service.update(1, productRequest)

        Assertions.assertNotNull(actual)
        Assertions.assertEquals(randomName, actual.name)
    }

    @Test
    fun `if updated product not exist should return ResourceNotFoundException`() {
        assertThrows<ResourceNotFoundException> {
            service.update(Long.MAX_VALUE, productRequest)
        }
    }

    @Test
    fun `if exist should delete product`() {
        val product = service.save(productRequest)
        val response = service.delete(product.id)

        Assertions.assertEquals("kotlin.Unit", response::class.qualifiedName)
    }

}

